File: go/packer.go
package packer

import (
	"encoding/binary"
	"fmt"
)

type Packer struct {
	buffer []byte
	length int
}

func NewPacker() *Packer {
	return &Packer{
		buffer: make([]byte, 0),
		length: 0,
	}
}

func (p *Packer) GetBuffer() []byte {
	return p.buffer
}

func (p *Packer) GetSize() int {
	return p.length
}

func (p *Packer) GetBufferWithLength() []byte {
	lengthBytes := make([]byte, 4)
	binary.LittleEndian.PutUint32(lengthBytes, uint32(p.length))
	return append(lengthBytes, p.buffer...)
}

func (p *Packer) encodeUTF8(data interface{}) []byte {
	if str, ok := data.(string); ok {
		return []byte(str)
	}
	if bytes, ok := data.([]byte); ok {
		return bytes
	}
	return nil
}

func (p *Packer) AddInt(data int32) {
	dataBytes := make([]byte, 4)
	binary.LittleEndian.PutUint32(dataBytes, uint32(data))
	p.buffer = append(p.buffer, dataBytes...)
	p.length += 4
}

func (p *Packer) AddStr(data string) {
	encodedData := p.encodeUTF8(data)
	lengthBytes := make([]byte, 4)
	binary.LittleEndian.PutUint32(lengthBytes, uint32(len(encodedData)+1))
	p.buffer = append(p.buffer, lengthBytes...)
	p.buffer = append(p.buffer, encodedData...)
	p.buffer = append(p.buffer, 0x00)
	p.length += len(lengthBytes) + len(encodedData) + 1
}

func (p *Packer) AddData(data []byte) {
	lengthBytes := make([]byte, 4)
	binary.LittleEndian.PutUint32(lengthBytes, uint32(len(data)))
	p.buffer = append(p.buffer, lengthBytes...)
	p.buffer = append(p.buffer, data...)
	p.length += len(lengthBytes) + len(data)
}

func (p *Packer) Dump() {
	fmt.Printf("[*] Buffer: [%d] [%x]\n", p.length, p.buffer)
}

File: Packer.py
#Code ported from https://github.com/HavocFramework/havoc-py

from struct import pack,calcsize, unpack

class Packer:

    def __init__(self):
        self.buffer = b''
        self.length = 0

    def get_buffer(self) -> bytes:
        return self.buffer

    def get_size(self) -> int:
        return self.length

    def get_buffer_with_length(self) -> bytes:
        return pack("<L", self.length ) + self.buffer

    def _encode_utf8(self, data: str) -> bytes:
        if isinstance(data, str):
            return data.encode("utf-8")
        return data

    def add_int(self, data : int ) -> None:
        self.buffer += pack("<i", data)
        self.length += 4
        return

    def add_str(self, data: str) -> None:
        encoded_data = self._encode_utf8(data)
        fmt = f"<L{len(encoded_data)}s"
        self.buffer += pack(fmt, len(encoded_data) + 1, encoded_data + b'\x00')
        self.length += calcsize(fmt)

    def add_data(self, data: bytes) -> None:
        fmt = f"<L{len(data)}s"
        self.buffer += pack(fmt, len(data), data)
        self.length += calcsize(fmt)
        
    def dump(self) -> None:
        print( f"[*] Buffer: [{ self.length }] [{ self.get_buffer() }]" )
        return


class Parser:
    buffer: bytes = b''
    length: int = 0

    def __init__(self, buffer, length):
        self.buffer = buffer
        self.length = length
        return

    def parse_int(self) -> int:
        val = unpack( ">i", self.buffer[ :4 ])
        self.buffer = self.buffer[4:]
        return val[ 0 ]

    def parse_bytes( self ) -> bytes:
        length      = self.parse_int()
        buf         = self.buffer[:length]
        self.buffer = self.buffer[length:]
        return buf

    def parse_str( self ) -> str:
        return self.parse_bytes().decode('utf-8')
